adding the ability to select different qualities for png, jpeg and we…

FailedSitcom opened the pull request:
…bp images

## Description

<!-- Write a brief description of the changes introduced by this PR -->

This gives the ability to set different qualities for different image formats:

```
query Example {
    resource {
      coverImage {
        fluid(maxWidth: 100, webp_quality: 90, jpeg_quality: 50, png_quality: 60) {
          ...GatsbyImageSharpFluid_withWebp
        }
      }
    }
  }
```

If none of the format specific qualities are passed, it will fallback to the `quality` argument if passed, otherwise the default quality will be used.

Before this is merged I would also like to add documentation explaining how it works, but as this is my first experience in contributing to Gatsby, I wanted to see if I've gone about this in the right way and if there are any concerns that I'm missing.

## Related Issues

Fixes https://github.com/gatsbyjs/gatsby/issues/10203

coreyward commented:
At a glance this looks pretty good, but I think it's important to also update the official Contentful-source plugin that integrates with Gatsby Image with these changes as well. 

https://github.com/gatsbyjs/gatsby/blob/2f4c4608df62e867f646881a00a5141860b7f5e0/packages/gatsby-source-contentful/src/extend-node-type.js#L62-L79

FailedSitcom commented on packages/gatsby-plugin-sharp/src/process-file.js:
```diff
@@ -211,10 +217,12 @@ const compressPng = (pipeline, outputPath, options) =>
+                  options.png_quality + 25,
+                  100
+                )}`
+              : `${options.quality}-${Math.min(options.quality + 25, 100)}`, // e.g. 40-65
```
@coreyward nice call

FailedSitcom commented:
Thank you for the feedback @coreyward! I saw your mention of the Contentful integration on the original issue, and wanted to look into it before requesting a further review (but am very pleased you were able to point me to the relevant part of the code).

If I implemented it this week, would you be interested in taking a look?

coreyward commented:
I'd be happy to help review, but I'm not actually that familiar with that part of the codebase. I'm confident that one of the members of the core team will review and provide additional guidance as well—they do a great job with new contributors.

FailedSitcom commented:
Well I appreciate the time you've already taken to look over it, and will definitely check-in with the core team when I've addressed the points you raised.

Thanks again!

vladar commented on packages/gatsby-plugin-sharp/README.md:
```diff
@@ -67,6 +70,9 @@ Automatically create sizes for different resolutions — we do 1x, 1.5x, and 2x.
 - `quality` (int, default: 50)
+- `jpeg_quality` (int)
+- `png_quality` (int)
+- `webp_quality` (int)
```
Maybe replace underscores with camelCase as this looks inconsistent now (i.e. `png_quality` but `pngCompressionLevel`)?

vladar replied:
We also need a note in docs about this feature. I guess the original issue description is a good start for it.

vladar commented on packages/gatsby-plugin-sharp/src/plugin-options.js:
```diff
@@ -11,6 +11,9 @@ const pluginDefaults = {
   quality: 50,
+  jpeg_quality: ``,
+  png_quality: ``,
+  webp_quality: ``,
```
Having those args as empty strings seem a bit weird. Maybe we could change them to `null` (or `undefined`) and make sure they are handled properly in the `healOptions` function (with `options.quality` as a fallback as suggested by @coreyward below).

vladar commented:
@FailedSitcom thanks for the PR - it will be beneficial for the community once merged.

Please let me know if I can help you move this forward?

This could be a good case for a [pair programming](https://www.gatsbyjs.org/contributing/pair-programming/) session. Or, if you don't have time, you could provide us push-access to your branch so that we could do those tweaks and merge?

FailedSitcom commented:
@vladar thank you for your helpful suggestions and encouragement. I'd sadly been too busy over the last couple of weeks to implement the changes you suggested, but would be more than happy to this weekend.

If something comes up and I'm unable to do so, I'd be more than happy to provide push-access.

Thank you (and the awesome Gatsby team) again!

vladar commented:
Sounds good! Thanks for doing this!
